scheduler: cleanup: simplify logic and cleanup

anandbonde opened the pull request:
Simplify scheduler and group logic, clearer naming and reduced syntactic clutter.

anandbonde committed:
scheduler: cleanup: simplify logic and cleanup

Simplify scheduler and group logic, clearer naming and reduced syntactic
clutter.

iyzhang commented on src/runtime/scheduler/page/waker_ref.rs:
```diff
@@ -92,17 +92,17 @@ impl Drop for WakerRef {
 unsafe fn waker_ref_clone(ptr: *const ()) -> RawWaker {
-    let p: WakerRef = WakerRef(NonNull::new_unchecked(ptr as *const u8 as *mut u8));
-    let q: WakerRef = p.clone();
+    let p = WakerRef(NonNull::new_unchecked(ptr as *const u8 as *mut u8));
```
Do you want to give this a full variable name?

iyzhang commented on src/runtime/scheduler/page/waker_ref.rs:
```diff
@@ -142,130 +142,121 @@ mod tests {
+        let page = WakerPageRef::default();
+        crate::ensure_eq!(page.refcount(), 1);
+
+        let q = WakerRef::new(page.as_raw_waker_ref(0));
```
same here? Something like raw_waker_ref or page_waker_ref

iyzhang commented on src/runtime/scheduler/group.rs:
```diff
@@ -51,170 +49,140 @@ pub struct TaskGroup {
-        waker_page_ref.clear(waker_page_offset);
-        if let Some(task) = self.tasks.remove_unpin(pin_slab_index) {
+    /// Remove task by handle.
+    pub fn remove(&mut self, idx: usize) -> Option<Box<dyn Task>> {
```
I feel like the difference between idx and i is no longer clear. We should at least include the data structures that each index into in the variable name.

iyzhang commented on src/runtime/scheduler/group.rs:
```diff
@@ -51,170 +49,140 @@ pub struct TaskGroup {
-    }
-
-    fn get_waker_page_offset(pin_slab_index: usize) -> usize {
+    fn get_page_offset(pin_slab_index: usize) -> usize {
```
should we get rid of the get_ on these function names as well and just name it page_offset?

anandbonde replied:
These are non-instance fns, so having `get_*` namesreduces ambiguity.

> e.g. `Self::page_offset(idx)` is confusing whether it's querying or setting the offset.

However, for instance fns it seems more natural to remove the prefix.

> e.g. `page.offset()`

I think we can leave them as they are, unless you have a preference.

iyzhang replied:
yes, this makes sense.

iyzhang commented on src/runtime/scheduler/group.rs:
```diff
@@ -51,170 +49,140 @@ pub struct TaskGroup {
-                Some(max_iterations) if iterations_run >= max_iterations => return completed_tasks,
-                _ => iterations_run += 1,
+
+            if let Some(limit) = limit {
```
this could be written as one line with is_some_and

iyzhang commented on src/runtime/scheduler/scheduler.rs:
```diff
@@ -55,40 +55,27 @@ impl Scheduler {
-    pub fn insert_and_poll_task<T: Task>(&mut self, group_id: SchedulerId, task: T) -> Option<InsertResult> {
-        let group: &mut TaskGroup = self.get_mut_group(group_id)?;
-        group.insert_and_poll_task(Box::new(task))
+    pub fn insert<T: Task>(&mut self, group_id: SchedulerId, task: T) -> Option<InsertResult> {
```
I think we might want to name this insert_and_run task, just to make it really clear that we are going to trigger the task right away.

anandbonde replied:
The "run" part is an implementation detail so I left it out. Another option is to invert the implementation detail responsibility and just call it `run` such that "insert" is just a side-effect of running.

We should evaluate how it may be perceived by the reader... Thoughts?

iyzhang replied:
yes, maybe we could call it schedule_task to indicate that its being added to the scheduler and scheduled to run.

anandbonde replied:
Changed all the related fns to be `schedule*` so that it flows naturally with the functionality.

iyzhang commented on src/runtime/scheduler/scheduler.rs:
```diff
@@ -55,40 +55,27 @@ impl Scheduler {
-        // Keep polling the group and checking for runnable tasks until there are none left.
-        group.poll_group(max_iterations, true)
+    /// Do not use this function on coroutines that use poll_yield, unless limit is set.
+    pub fn poll_group(&mut self, id: SchedulerId, limit: Option<usize>) -> Vec<Box<dyn Task>> {
```
we should probably settle on either poll or run as the verb in all of our variable names

anandbonde replied:
I've changed it, let me know what you think.

iyzhang replied:
Great

anandbonde committed:
scheduler: rename insert to schedule and related fns

We decided that `schedule` is more expressive than `insert` when adding
a task to the scheduler. So renamed all the related fn chain in the
codebase.
